### ISSUE TITLE
Improve error locations for return type errors

### ISSUE THREAD

**Issue by Andarist** [microsoft/TypeScript#52934]
# Suggestion

## 🔍 Search Terms

error return type location squiggles

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

I propose to move the error location for return type error to the `return` keyword of the return statement.

At the moment the whole return statement gets underlined. In almost every case that is longer than 1 LOC it feels very overwhelming. At the moment I'm looking at this:
<img alt="multiline expression underlined with red squiggles in VS Code" width="483" alt="Screenshot 2023-02-23 at 14 56 18" src="https://user-images.githubusercontent.com/9800850/220928096-5859a6ac-04e9-4165-a7c5-763c763a6b29.png">

It makes it hard to focus on what is actually underlined here - it's not even immediately obvious that this complains about the type of this expression not being assignable to the declared return type. The first instinct tell me to look into the expressions itself (since it's a call expression) and look for some errors in it - it doesn't exactly point my mind to thinking about the failed assignability check with the declared return type. And even when I finally recognize the true problem - it's distracting and it makes it harder (for me) to focus on the problem.

If we compare this with assignability errors reported on variable declarations it also feels somewhat inconsistent:
```ts
const a: string = 10
//    ^
function test(): string {
    return 10
//  ^^^^^^^^^    
}
```

With the variable declaration only the "target" of the assignment gets highlighted. That target is almost guaranteed to be short(ish) as that's just an identifier. So this error just can't span multiple lines.

What I'm proposing is to change the error location to this:
```ts
function test(): string {
    return 10
//  ^^^^^^    
}
```

cc @DanielRosenwasser , i think i recall some similar issues created by you (but perhaps about other locations) but I can't find them now

**Comment by jakebailey** [comment]
I feel like sticking it onto the return keyword is fine, if only because some cases of error locations for non-assignability also use a smaller range: [Playground Link](https://www.typescriptlang.org/play?#code/PTAECUEMDsHMFNQEsDOoAuALR9rqQE6JHoCuB0oK6k68AtrugHQBQAZqdAMb4D2ldnz4AKAJQAuUNFL0ARvAKgA3q1DrQJcpQBEkFABN2OgNysAvq1YgIMBMjQArUtQzZQAN0gEkkOQBt4Nm4BVy9-UngpGXlFUABeUD1DYzNrMCg4RFRQZ1csRAAHAj5CxXQAT2lIRmDQ9FAADyllUGLS6NkFJXMElTaSwqlkox1QSysbTPscgtAmQkRwyLYDeG5-b0ROHn5KF3gAOS7FEWWo6ROCSU8+JAMzA+PYghEdAEYAJgBmABYdMRpdK2LIOXIuBpzdplAiVaq1VhrDZbUA7XhIASgA4AeTkjjOkAiF1a0M6L3GNw8dwerBxeJEJMGw30owpaRsAFFGtx4IUGgB3bCUJANVDQADkLFpKHguPxI2MgKAA)

**Comment by fatcerberus** [comment]
It looks like the general rule is that the error span is always on the innermost offending assignment target (object property, variable, etc.), except when that's impossible because the assignment target isn't explicitly part of syntax, e.g. function parameters, in which case it highlights the source value.  `return` is a notable outlier in that it highlights *both* the assignment target (`return`) *and* the source.

**Comment by DanielRosenwasser** [comment]
Thanks @Andarist!
